synapse/handlers/presence.py

```diff
@@ -967,9 +988,24 @@ async def user_syncing(
             self.user_to_num_current_syncs[user_id] = curr_sync + 1
 
             prev_state = await self.current_state_for_user(user_id)
+
+            # If they're busy then they don't stop being busy just by syncing,
+            # so just update the last sync time.
+            if prev_state.state != PresenceState.BUSY:
+                # XXX: We set_state separately here and just update the last_active_ts above
+                # This keeps the logic as similar as possible between the worker and single
+                # process modes. Using set_state will actually cause last_active_ts to be
+                # updated always, which is not what the spec calls for, but synapse has done
+                # this for... forever, I think.
+                await self.set_state(
+                    UserID.from_string(user_id), {"presence": presence_state}, True
+                )
+                # To keep the single process behaviour consistent with worker mode, run the
+                # same logic as `update_external_syncs_row`, even though it looks weird
+                # (this should come from the in-memory cache)
+                prev_state = await self.current_state_for_user(user_id)
+
```

Comment:
Sorry if I wasn't very clear just now. The weirdness (in my mind at least) is calling `set_state` and then replacing an `OFFLINE` state that we've just set with `ONLINE` (not that it can happen). It's crying out for a refactor which we can't do without mucking with the inter-worker API.

```suggestion
                # Retrieve the new state for the logic below. This should come from the
                # in-memory cache.
                prev_state = await self.current_state_for_user(user_id)

            # To keep the single process behaviour consistent with worker mode, run the
            # same logic as `update_external_syncs_row`, even though it looks weird.
```

Reply:
Ah, yes - I see what you mean.